src/pam_hbac_ldap_compat.c

```diff
@@ -0,0 +1,284 @@
+static int rdn_keyval(const char *exploded_rdn_value,
+                      char **_key, char **_val)
+{
+    size_t i;
+    char *eq_sign;
+    const char *rdn_start;
+    const char *rdn_end;
+    const char *val_start;
+    char *key = NULL;
+    char *value = NULL;
+
+    if (exploded_rdn_value == NULL) {
+        return EINVAL;
+    }
+
+    /* Skip leading whitespace before the key */
+    rdn_start = exploded_rdn_value;
+    for (i = 0; isspace(rdn_start[i]); i++);
+
+    rdn_start += i;
+    if (*rdn_start == '\0') {
+        return EINVAL;
+    }
+
+    eq_sign = strchr(rdn_start, '=');
+    if (eq_sign == NULL) {
+        return EINVAL;
+    }
+
+    /* Skip whitespace before the '=' sign and after the key */
+    for (rdn_end = eq_sign;
+         rdn_end > rdn_start && isspace(*rdn_end);
+         rdn_end--);
+
+    val_start = eq_sign+1;
+    /* Skip leading whitespace before the value */
+    for (i =0; isspace(val_start[i]); i++);
+
+    /* Empty value? */
+    val_start += i;
+    if (*val_start == '\0') {
+        return EINVAL;
+    }
+
+    key = strndup(rdn_start, rdn_end - rdn_start);
+    value = strdup(val_start);
+    if (key == NULL || value == NULL) {
+        free(key);
+        free(value);
+        return ENOMEM;
+    }
+
+    *_key = key;
+    *_val = value;
+    return 0;
+}
+
+static LDAPAVA *rdn2ava(const char *rdn)
+{
+    char **exploded_rdn = NULL;
+    LDAPAVA *ava;
+    char *key = NULL;
+    char *val = NULL;
+    int ret;
+
+    /* For single-valued RDNs, this is a noop, for multiple-valued, we just
+     * hope for the best. Because we don't support deny rules, the user
+     * might be not allowed access at worst
+     */
+    exploded_rdn = ldap_explode_rdn(rdn, 0);
+    if (exploded_rdn == NULL) {
+        return NULL;
+    }
+
+    ava = malloc(sizeof(LDAPAVA));
+    if (ava == NULL) {
+        ldap_value_free(exploded_rdn);
+        return NULL;
+    }
+
+    /* Since the server is IPA, we can ignore anything but the simplest,
+     * single-valued RDNs, we know IPA doesn't use those
+     */
+    ret = rdn_keyval(exploded_rdn[0], &key, &val);
+    ldap_value_free(exploded_rdn);
+    if (ret != 0 || key == NULL || val == NULL) {
+        free(key);
+        free(val);
+        free(ava);
+        return NULL;
+    }
+
+    ava->la_attr.bv_val = key;
+    ava->la_value.bv_val = val;
+    ava->la_attr.bv_len = strlen(key);
+    ava->la_value.bv_len = strlen(val);
+
+    return ava;
+}
+
+static int str2rdn(const char *str_rdn, LDAPRDN *_rdn)
+{
+    LDAPAVA **new_rdn;
+
+    /* For single-valued RDNs, this is a noop, for multiple-valued, we just
+     * hope for the best. Because we don't support deny rules, the user
+     * might be not allowed access at worst
+     */
+    new_rdn = calloc(2, sizeof(LDAPAVA *));
+    new_rdn[0] = rdn2ava(str_rdn);
+    if (new_rdn[0] == NULL) {
+        free(new_rdn);
+        return ENOMEM;
+    }
+    new_rdn[1] = NULL;
```

Comment:
On Mon, May 23, 2016 at 07:49:16AM -0700, Pavel Reichl wrote:
> > +
> > +static int str2rdn(const char *str_rdn, LDAPRDN *_rdn)
> > +{
> > +    LDAPAVA **new_rdn;
> > +
> > +    /* For single-valued RDNs, this is a noop, for multiple-valued, we just
> > +     * hope for the best. Because we don't support deny rules, the user
> > +     * might be not allowed access at worst
> > +     */
> > +    new_rdn = calloc(2, sizeof(LDAPAVA *));
> > +    new_rdn[0] = rdn2ava(str_rdn);
> > +    if (new_rdn[0] == NULL) {
> > +        free(new_rdn);
> > +        return ENOMEM;
> > +    }
> > +    new_rdn[1] = NULL;
> 
> I don't think this line is necessary as calloc() should initialize memory to NULL, but I don't mind to do it explicitly... 

Right, removed.